playkit/src/main/java/com/kaltura/playkit/player/TrackSelectionHelper.java

```diff
@@ -254,10 +300,92 @@ private String getLanguageFromFormat(Format format) {
         return format.language;
     }
 
+    private List<VideoTrack> filterVideoTracks() {
+
+        if (videoTracksCodecsMap == null || videoTracksCodecsMap.isEmpty()) {
+            return videoTracks;
+        }
+
+        boolean atLeastOneCodecSupportedInHardware = videoCodecsSupportedInHardware(); // if no hardware decoders play with the software
+
+        VideoCodecSettings preferredVideoCodecSettings = playerSettings.getPreferredVideoCodecSettings();
+        if (preferredVideoCodecSettings.getAllowVideoMixedMimeTypeAdaptiveness()) {
+            populateAllCodecTracks(atLeastOneCodecSupportedInHardware);
+            return videoTracks;
+        }
+
+        for (PKVideoCodec videoCodecForPlayback : preferredVideoCodecSettings.getVideoCodecPriorityList()) {
+            if (!videoTracksCodecsMap.containsKey(videoCodecForPlayback)) {
+                continue;
+            }
+
+            if (videoTracksCodecsMap.get(videoCodecForPlayback) != null && videoTracksCodecsMap.get(videoCodecForPlayback).isEmpty()) {
+                continue;
+            }
+
+            VideoTrack candidateVideoTrack = videoTracksCodecsMap.get(videoCodecForPlayback).get(0);
+            if (candidateVideoTrack.getCodecName() != null &&
+                    isCodecSupported(candidateVideoTrack.getCodecName(), TrackType.VIDEO, false)) {
+                return videoTracksCodecsMap.get(videoCodecForPlayback);
+            } else if ((!atLeastOneCodecSupportedInHardware  || preferredVideoCodecSettings.isAllowSoftwareDecoder()) &&
+                    (candidateVideoTrack.getCodecName() != null && isCodecSupported(candidateVideoTrack.getCodecName(), TrackType.VIDEO, true))) {
+                return videoTracksCodecsMap.get(videoCodecForPlayback);
+            }
+        }
+
+        if (videoTracksCodecsMap.containsKey(PKVideoCodec.HEVC)) {
+            return  videoTracksCodecsMap.get(PKVideoCodec.HEVC);
+        } else if (videoTracksCodecsMap.containsKey(PKVideoCodec.AV1)) {
+            return  videoTracksCodecsMap.get(PKVideoCodec.AV1);
+        } else if (videoTracksCodecsMap.containsKey(PKVideoCodec.VP9)) {
+            return  videoTracksCodecsMap.get(PKVideoCodec.VP9);
+        } else if (videoTracksCodecsMap.containsKey(PKVideoCodec.VP8)) {
+            return  videoTracksCodecsMap.get(PKVideoCodec.VP8);
+        } else if (videoTracksCodecsMap.containsKey(PKVideoCodec.AVC)) {
+            return  videoTracksCodecsMap.get(PKVideoCodec.AVC);
+        } else {
+            return videoTracks;
+        }
```

Comment:
For loop - create local array with all codecs by priority.

Reply:
Converted to forloop